plugins/inputs/mqtt_consumer/mqtt_consumer.go

```diff
@@ -240,6 +245,36 @@ func compareTopics(expected []string, incoming []string) bool {
 }
 
 func (m *MQTTConsumer) onMessage(acc telegraf.TrackingAccumulator, msg mqtt.Message) error {
+	var remainingLength int
+	var qosFlagsSize int
+	var qOsSize int
+	topicSize := len(msg.Topic()) + 2
+	switch int(msg.Qos()) {
+	case 1:
+		qOsSize = 2
+		qosFlagsSize = 4
+	case 2:
+		qOsSize = 2
+		qosFlagsSize = 12
+	default:
+		qOsSize = 0
+	}
+	payloadSize := len(msg.Payload())
+	remainingContent := topicSize + qOsSize + payloadSize
+	if remainingContent < 1<<7 {
+		remainingLength = 1
+	} else if remainingContent < 1<<14 {
+		remainingLength = 2
+	} else if remainingContent < 1<<21 {
+		remainingLength = 3
+	} else {
+		remainingLength = 4
+	}
+	publishMessageSize := remainingLength + remainingContent + 1
+	byteCount := publishMessageSize + qosFlagsSize
```

Comment:
Thanks for these changes. They will be useful!

Should this only be reporting the payload size instead of the whole message size? The message has been parsed by now.

The rest of the code here seems to be making an educated guess on what the actual message size was at a lower level. I'm concerned that this isn't correct and there's no way to easily check it. I also wonder if the mqtt module will change and make this inaccurate in the future.

Is there a way to get the message size directly from eclipse/paho.mqtt.golang? If not, could we look into adding it upstream?

Reply:
Since the message has already been parsed, what I did to get the total incoming message size was work backwards from the given topic, payload, and flags. I don't want just the payload size, I want to get the total incoming packet size in order to know how much data usage a client is using. I was recommended to go to the eclipse/paho.mqtt.golang community to see if there was a way to get the message size directly from there, and the technique I implemented here was based off of their suggestion [here](https://groups.google.com/g/mqtt/c/Bk0ajGos-Sw). I suppose it could be possible to add this metric directly to eclipse/paho.mqtt.golang but what would be the advantages of doing it there rather than here?

Reply:
A message in that thread illustrates my concern:

 > ... and if it's MQTT v5 then you need to add some more on to account for properties, but you won't absolutely get all the information that you need to correctly calculate the size in a user library."

The calculations here may be accurate for your use case, but evidently they're not accurate for all use cases (at least not v5) and it may not be possible to know when they're accurate.

I do think the payload size and number of messages are great internal metrics because they're directly measured, not derived.

Reply:
I see what you mean. So would you recommend I change it to just collect the payload size and number of messages metrics rather than trying to get the total size of the message?

Reply:
I don't see anything preventing this from being merged if we remove the code to calculate the total size. Maybe we should go ahead with that, then if/when we find a way to directly measure the total size we can add it in another PR.